### PR TITLE
Use an INSERT rather than an UPSERT when only touching partition/clustering columns

### PR DESCRIPTION
**Pull request by suhailpatel**
This is protection [against a nasty Cassandra bug](https://issues.apache.org/jira/browse/CASSANDRA-11805) where UPSERTing a row with all the non-partition/clustering column keys as 'null' causes the entire row to be deleted. This is undesirable behaviour and we should rather use an INSERT to retain the original data

### CODE REVIEW

File: table.go [hunk 1]
```diff
@@ -161,10 +181,8 @@ func (t t) Set(i interface{}) Op {
...
 	ks := append(t.info.keys.PartitionKeys, t.info.keys.ClusteringColumns...)
 	updFields := removeFields(m, ks)
-	if len(updFields) == 0 {
-		return newWriteOp(t.keySpace.qe, filter{
-			t: t,
-		}, insertOpType, m)
+	if len(updFields) == 0 || allFieldValuesAreNullable(updFields) {
+		return newWriteOp(t.keySpace.qe, filter{t: t}, insertOpType, m)
 	}
 	transformFields(updFields)
```
Comment (mattheath) [inline table.go:187]: This `transformFields` appears to just handle counter increments?

File: table.go [hunk 2]
```diff
@@ -124,6 +125,25 @@ func transformFields(m map[string]interface{}) {
...
 
+// allFieldValuesAreNullable checks if all the fields passed in contain
+// items that will eventually be marked in Cassandra as null (so we can
+// better predict if we should do an INSERT or an UPSERT). This is to
+// protect against https://issues.apache.org/jira/browse/CASSANDRA-11805
+func allFieldValuesAreNullable(fields map[string]interface{}) bool {
+	for _, value := range fields {
+		rv := reflect.ValueOf(value)
+		switch rv.Kind() {
+		case reflect.Array, reflect.Slice, reflect.Map:
```
Comment (mattheath) [inline table.go:136]: Just to check — are these the only types which could be nullable? 

I'm trying to see where passing an incorrect type would be filtered out, but I can't see anything that does this? Is it possible for value to be a nil pointer for example?

In `Set()` below ([L177](https://github.com/monzo/gocassa/pull/60/files#diff-cce950b40a3962c50522cd22e55b0b51R177)) a map[string]interface{} with a pointer type (say a nil pointer to a string — `(*string)(nil)`) seems to be passed through and would result in newWriteOp returning an `updateOpType`?

https://play.golang.org/p/lOA15Bdk5v3

Would this be filtered out later by GoCQL as a field of type `*string` would be invalid, or converted to type `string` and therefore `nil -> ""` ?

File: table.go [hunk 3]
```diff
@@ -124,6 +125,25 @@ func transformFields(m map[string]interface{}) {
 	}
 }
 
+// allFieldValuesAreNullable checks if all the fields passed in contain
+// items that will eventually be marked in Cassandra as null (so we can
+// better predict if we should do an INSERT or an UPSERT). This is to
```
[the remark this answers is not in the document]
Comment (suhailpatel) [reply table.go:130]: I specifically wanted upsert to describe this behaviour. The intention is not to map to a cql command but rather a behaviour of the library (which is to do an upsert)

File: table.go (hunk 2 above)
Comment (suhailpatel) [reply table.go:136]: Pointers are filtered out a layer above through `reflect.Indirect` when the [struct is converted into a map](https://github.com/monzo/gocassa/blob/master/reflect/reflect.go) to begin with. 

Also yes, those are the fields that I found to be susceptible to the nullable behaviour.

Comment (suhailpatel) [reply table.go:136]: I have [added an explicit test](https://github.com/monzo/gocassa/pull/60/commits/d73135a970d51b27a12c25a28a36383cc3f0cbd7) for this behaviour just to make sure it's codified

**Review by mattheath — APPROVED** [review]
Looks great, thanks for adding the really thorough tests! ❤️

**Pull request merged**